### PR TITLE
Get health check path from readiness probe

### PR DESCRIPTION
**Pull request by glindste**
This PR is not ready but I wanted to open it because I would like some feedback on the approach before continuing. I tried reusing as much as possible from `ingress-gce`, however I chose to copy `GetProbe` and `getHTTPProbe` from `ingress-gce/controller/utils.go`. This is because the `GCETranslator` that was the receiver needed a `LoadBalancerController` to initialize which seemed like unnecessary bloat. One possible solution would be to refactor the code in `ingress-gce` to make it easier to reuse so we could avoid code duplication. I haven't bothered with the tests yet since I'm not sure if this is the way you'd want to implement this.

I tested the implementation manually by running and it works ~with one exception, sometimes it seems like even though the cached informer has been synced it fails to get the correct path and defaults to `/`. I'm not yet sure if there's an issue with some race condition in the syncing or if the error is in this PR.~ EDIT: found the issue, I used the wrong `HasSynced` method when checking if the pod informer had synced

### CODE REVIEW

**Comment by G-Harmon** [comment]
cc @nikhiljindal @csbell @bowei

**Comment by nikhiljindal** [comment]
/ok-to-test

Hey thanks a lot for the PR @glindste. Appreciate the contribution. 

Regarding code duplication with `ingress-gce`, we are working with @bowei to reduce it. We have existing copied code here and want to get rid of it as soon as possible, but realize that requires refactoring in ingress-gce and then revendoring it here.

I am curious how you found this problem? Are you using kubemci?

**Comment by glindste** [comment]
@nikhiljindal I spent last week researching how to do multi cluster ingress. Kubernetes federation did not fit our usecase at all, we want individual control over the clusters and it doesn't mesh well with how we deploy things. I think I was digging around in the `ingress-gce` source code and found the `gce-multi-cluster` annotation which led me to kubemci, which seems to do exactly what we want.

Getting the health check path from the probe was pretty much the only missing feature before I could test how kubemci worked (neither our service or the default backend returns 200 on `/`) so I figured I'd give it a go.

File: app/kubemci/pkg/gcp/healthcheck/healthchecksyncer.go [hunk 1]
```diff
@@ -210,7 +231,7 @@ func (h *HealthCheckSyncer) desiredHealthCheck(lbName string, port ingressbe.Ser
 	case "HTTP":
 		hc.HttpHealthCheck = &compute.HTTPHealthCheck{
 			Port:        port.Port,
-			RequestPath: "/", // TODO(nikhiljindal): Allow customization.
+			RequestPath: path, // TODO(nikhiljindal): Allow customization.
```
Comment (nikhiljindal) [inline app/kubemci/pkg/gcp/healthcheck/healthchecksyncer.go]: We can remove the TODO now :)

File: app/kubemci/pkg/gcp/healthcheck/healthchecksyncer.go [hunk 2]
```diff
@@ -66,12 +69,30 @@ var _ HealthCheckSyncerInterface = &HealthCheckSyncer{}
...
 // Returns a map of the ensured health checks keyed by the corresponding port.
-func (h *HealthCheckSyncer) EnsureHealthCheck(lbName string, ports []ingressbe.ServicePort, forceUpdate bool) (HealthChecksMap, error) {
+func (h *HealthCheckSyncer) EnsureHealthCheck(lbName string, ports []ingressbe.ServicePort, client kubernetes.Interface, forceUpdate bool) (HealthChecksMap, error) {
 	fmt.Println("Ensuring health checks")
 	var err error
 	ensuredHealthChecks := HealthChecksMap{}
 	for _, p := range ports {
-		hc, hcErr := h.ensureHealthCheck(lbName, p, forceUpdate)
+		ctx := ingresscx.NewControllerContext(client, p.SvcName.Namespace, 1*time.Second, false)
+		ki := kubeutils.NewKubernetesInformer(ctx)
```
Comment (nikhiljindal) [inline app/kubemci/pkg/gcp/healthcheck/healthchecksyncer.go]: We dont need the full informer. Informer watches all services and pods and is a good model for controller that syncs multiple ingresses (all ingresses at some point). We use the informer to optimize such that we fetch resources only once and keep them in memory until we get a watch notification for change.

Here we are concerned with a single ingress, so we can fetch the service and pods directly for that ingress only. We know the services here. We can fetch pods corresponding to the services and check for their readiness probe. Since the health check is same for all, all pods are required to have the same readiness probe.

File: app/kubemci/pkg/kubeutils/utils.go [hunk 3]
```diff
@@ -93,3 +103,120 @@ var getClientset = func(kubeconfigPath, contextName string) (kubeclient.Interfac
...
+	var svcPort api_v1.ServicePort
+	var found bool
+OuterLoop:
+	for _, as := range sl {
+		service = *as.(*api_v1.Service)
+		for _, sp := range service.Spec.Ports {
+			svcPort = sp
+			// only one Service can match this nodePort, try and look up
+			// the readiness probe of the pods behind it
+			if int32(port.Port) == sp.NodePort {
```
Comment (nikhiljindal) [inline app/kubemci/pkg/kubeutils/utils.go]: We have already done this part while computing ServicePort: https://github.com/GoogleCloudPlatform/k8s-multicluster-ingress/blob/ecff59d1b6493bc28bab02cb4de74c23e9b047b6/app/kubemci/pkg/gcp/loadbalancer/loadbalancersyncer.go#L365
Here we can directly fetch service with name port.SvcName to get the required service.

File: app/kubemci/pkg/kubeutils/utils.go [hunk 4]
```diff
@@ -93,3 +103,120 @@ var getClientset = func(kubeconfigPath, contextName string) (kubeclient.Interfac
...
+}
+
+// geHTTPProbe returns the http readiness probe from the first container
+// that matches targetPort, from the set of pods matching the given labels.
+func (t *KubernetesInformer) getHTTPProbe(svc api_v1.Service, targetPort intstr.IntOrString, protocol ingressut.AppProtocol) (*api_v1.Probe, error) {
+	l := svc.Spec.Selector
+
+	// Lookup any container with a matching targetPort from the set of pods
+	// with a matching label selector.
+	pl, err := t.podLister.List(labels.SelectorFromSet(labels.Set(l)))
```
Comment (nikhiljindal) [inline app/kubemci/pkg/kubeutils/utils.go]: Here we are fetching the pods for the given service. Instead of fetching them from podLister, use go-client directly and fetch them from apiserver.

File: app/kubemci/pkg/kubeutils/utils.go [hunk 5]
```diff
@@ -93,3 +103,120 @@ var getClientset = func(kubeconfigPath, contextName string) (kubeclient.Interfac
...
+	if err != nil {
+		return nil, err
+	}
+
+	// If multiple endpoints have different health checks, take the first
+	sort.Sort(ingressct.PodsByCreationTimestamp(pl))
```
Comment (nikhiljindal) [inline app/kubemci/pkg/kubeutils/utils.go]: We should generate an error if pods have different healthchecks

File: app/kubemci/pkg/kubeutils/utils.go [hunk 6]
```diff
@@ -23,6 +23,16 @@ import (
 	multierror "github.com/hashicorp/go-multierror"
 	kubeclient "k8s.io/client-go/kubernetes"
 	"k8s.io/client-go/tools/clientcmd"
+	"sort"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/apimachinery/pkg/util/intstr"
```
Comment (nikhiljindal) [inline app/kubemci/pkg/kubeutils/utils.go]: General FYI about ordering of imports.
We keep 3 sections: standard libs; non-standard libs outside this repo; libs from this repo.
sections are separated using a line break and imports are sorted within each section.

File: app/kubemci/pkg/gcp/loadbalancer/loadbalancersyncer.go [hunk 7]
```diff
@@ -113,7 +113,7 @@ func (l *LoadBalancerSyncer) CreateLoadBalancer(ing *v1beta1.Ingress, forceUpdat
 		return err
 	}
 	// Create health checks to be used by the backend service.
-	healthChecks, hcErr := l.hcs.EnsureHealthCheck(l.lbName, ports, forceUpdate)
+	healthChecks, hcErr := l.hcs.EnsureHealthCheck(l.lbName, ports, client, forceUpdate)
```
Comment (nikhiljindal) [inline app/kubemci/pkg/gcp/loadbalancer/loadbalancersyncer.go]: Note that you are passing go-client for only one cluster. You probably want to send the clients for all clusters.
To make it easier, you can break the changes into 2 stages:
* Make the happy path work: Compute the readiness probe for any one pod in any one cluster and use it in health check. This is easier, you need to pass only one client and fetch any one random pod and use its readiness probe. This works fine for the valid case where readiness probe is same for pods in all clusters.
* Detect the error scenario: Find the readiness probes of all pods for all services in all clusters and generate an error if they dont match. For this, you need to pass all clients, iterate over all of them to fetch pods in all clusters and then compare the readiness probes to generate an error if required. This is to handle error situations gracefully. We should detect the error scenario and generate a clear actionable error message for the user rather than showing them success and then they having to debug why is this not working.

You can directly do both 1 and 2 in this PR or just 1 for now in this PR. Let me know what you are comfortable with.

**Comment by nikhiljindal** [comment]
Thats great @glindste! Are you going to use this with your GCP clusters?

We'd like to include you (and others) in our early access program if possible. Could you sign up for the EAP using this form?
https://goo.gl/forms/RHg032DrRnnqF0by1 .
We would love your feedback as you try out kubemci and would love to know what works and what doesnt.
cc @mdelio

File: app/kubemci/pkg/gcp/loadbalancer/loadbalancersyncer.go (hunk 7 above)
Comment (glindste) [reply app/kubemci/pkg/gcp/loadbalancer/loadbalancersyncer.go]: Alright, I think I'll aim for 1 for now and see how it goes. If I have time to work on the next step when that's done I'll just continue here, otherwise I'll wrap up the PR and create an issue for step 2.

File: app/kubemci/pkg/gcp/healthcheck/healthchecksyncer.go (hunk 2 above)
Comment (glindste) [reply app/kubemci/pkg/gcp/healthcheck/healthchecksyncer.go]: 👍 Yeah I was mostly concerned with reusing as much code as possible from ingress-gce, but you're right it's better to keep it simple here and just fetch directly.

File: app/kubemci/pkg/gcp/loadbalancer/loadbalancersyncer.go (hunk 7 above)
Comment (nikhiljindal) [reply app/kubemci/pkg/gcp/loadbalancer/loadbalancersyncer.go]: sounds good. You can add a TODO for second to do it in another PR, if you dont get to it in this PR.

File: app/kubemci/pkg/kubeutils/utils.go (hunk 3 above)
Comment (glindste) [reply app/kubemci/pkg/kubeutils/utils.go]: Ah cool, thanks!

**Comment by glindste** [comment]
Thanks a lot for the feedback! :)

Yes if no major blocker comes up we'll probably use kubemci to set up the LB for our clusters, the alternative currently is managing the LB manually which we'd like to avoid. I've signed up for the EAP

File: app/kubemci/pkg/gcp/healthcheck/healthchecksyncer_test.go [hunk 8]
```diff
@@ -83,12 +102,22 @@ func TestEnsureHealthCheck(t *testing.T) {
 	}
 
 	for _, c := range testCases {
+		var svcPort int32 = 80
```
Comment (nikhiljindal) [inline app/kubemci/pkg/gcp/healthcheck/healthchecksyncer_test.go]: svcPort := 80 will also work?

File: app/kubemci/pkg/gcp/healthcheck/healthchecksyncer_test.go [hunk 9]
```diff
@@ -144,7 +238,7 @@ func TestDeleteHealthCheck(t *testing.T) {
 			Protocol: utils.ProtocolHTTP,
 		},
 	}
-	if _, err := hcs.EnsureHealthCheck(lbName, ports, clientfake.NewSimpleClientset(), false); err != nil {
+	if _, err := hcs.EnsureHealthCheck(lbName, ports, &clientfake.Clientset{}, false); err != nil {
```
Comment (nikhiljindal) [inline app/kubemci/pkg/gcp/healthcheck/healthchecksyncer_test.go]: Its not clear to me, honestly, when should I use NewSimpleClientset and when should I instantiate Clientset directly.

Curious to know your motivation for this change. Is this change required for the test to pass?

File: app/kubemci/pkg/kubeutils/utils.go [hunk 10]
```diff
@@ -216,7 +147,18 @@ func (t *KubernetesInformer) getHTTPProbe(svc api_v1.Service, targetPort intstr.
...
 		fmt.Printf("%v: lacks a matching HTTP probe for use in health checks.\n", logStr)
 	}
+
 	return nil, nil
 }
+
+// isSimpleHTTPProbe returns true if the given Probe is:
+// - an HTTPGet probe, as opposed to a tcp or exec probe
+// - has no special host or headers fields, except for possibly an HTTP Host header
+func isSimpleHTTPProbe(probe *api_v1.Probe) bool {
```
Comment (nikhiljindal) [inline app/kubemci/pkg/kubeutils/utils.go]: This is copied from ingress-gce?
Add a comment saying so and add a TODO to refactor to share rather than duplicate.

cc @bowei

File: app/kubemci/pkg/kubeutils/utils.go [hunk 11]
```diff
@@ -104,102 +99,38 @@ var getClientset = func(kubeconfigPath, contextName string) (kubeclient.Interfac
...
+		fmt.Printf("Unable to find pods backing service %v in namespace %v\n", port.SvcName.Name, port.SvcName.Namespace)
 		return nil, err
 	}
 
-	// If multiple endpoints have different health checks, take the first
-	sort.Sort(ingressct.PodsByCreationTimestamp(pl))
+	for _, pod := range pl.Items {
+		logStr := fmt.Sprintf("Pod %v matching service selectors %v (targetport %+v)", pod.Name, selector.String(), port.SvcTargetPort)
 
-	for _, pod := range pl {
```
Comment (nikhiljindal) [inline app/kubemci/pkg/kubeutils/utils.go]: Same: Note the place where this is copied from and add a TODO to refactor to share and not duplicate.
We should be able to refactor this code to extract the readiness probe, given the pods.

cc @bowei

File: app/kubemci/pkg/gcp/healthcheck/healthchecksyncer.go [hunk 12]
```diff
@@ -100,6 +91,37 @@ func (h *HealthCheckSyncer) EnsureHealthCheck(lbName string, ports []ingressbe.S
...
+		if probe != nil && probe.HTTPGet != nil {
+			if path != "" {
+				// Ensure that all probes have the same health check path
+				if path != probe.HTTPGet.Path {
+					pathErr := fmt.Errorf("probe path '%s' from client '%s' does not match already extracted health check path '%s'", probe.HTTPGet.Path, cName, path)
+					err = multierror.Append(err, pathErr)
+					return path, err
+				}
+			} else {
+				// TODO use same method as in ingress-gce/backend to get probe settings, which is a private method currently
```
Comment (nikhiljindal) [inline app/kubemci/pkg/gcp/healthcheck/healthchecksyncer.go]: Name the method here

File: app/kubemci/pkg/gcp/healthcheck/healthchecksyncer.go [hunk 13]
```diff
@@ -100,6 +91,37 @@ func (h *HealthCheckSyncer) EnsureHealthCheck(lbName string, ports []ingressbe.S
...
+		if prErr != nil {
+			prErr = fmt.Errorf("Error %s when getting probe for port %v", prErr, port)
+			err = multierror.Append(err, prErr)
+			continue
+		}
+		if probe != nil && probe.HTTPGet != nil {
+			if path != "" {
+				// Ensure that all probes have the same health check path
+				if path != probe.HTTPGet.Path {
+					pathErr := fmt.Errorf("probe path '%s' from client '%s' does not match already extracted health check path '%s'", probe.HTTPGet.Path, cName, path)
```
Comment (nikhiljindal) [inline app/kubemci/pkg/gcp/healthcheck/healthchecksyncer.go]: change "client %s" to "cluster %s". its the cluster name extracted from kubeconfig.
Also same comment as above about making this error msg more user friendly.  say readiness probe from pod spec

File: app/kubemci/pkg/gcp/healthcheck/healthchecksyncer.go [hunk 14]
```diff
@@ -100,6 +91,37 @@ func (h *HealthCheckSyncer) EnsureHealthCheck(lbName string, ports []ingressbe.S
 	return ensuredHealthChecks, err
 }
 
+func getHealthCheckPath(port ingressbe.ServicePort, clients map[string]kubernetes.Interface) (string, error) {
+	var err error
+	path := ""
+	for cName, c := range clients {
+		probe, prErr := kubeutils.GetProbe(c, port)
+		if prErr != nil {
+			prErr = fmt.Errorf("Error %s when getting probe for port %v", prErr, port)
```
Comment (nikhiljindal) [inline app/kubemci/pkg/gcp/healthcheck/healthchecksyncer.go]: "getting probe for port" is not really clear to the user.
How about saying "getting readiness probe from pods for service \<service-name\>?"

File: app/kubemci/pkg/gcp/healthcheck/healthchecksyncer_test.go [hunk 15]
```diff
@@ -35,65 +36,89 @@ import (
...
 		// Human-readable description of test.
 		desc string
 		// Inputs
 		protocol    utils.AppProtocol
 		forceUpdate bool
-		withProbe   bool
+		clients     map[string]kubernetes.Interface
 		// Outputs
-		ensureErr bool
+		hcPath      string
```
Comment (nikhiljindal) [inline app/kubemci/pkg/gcp/healthcheck/healthchecksyncer_test.go]: expectedHCPath?

**Comment by nikhiljindal** [comment]
Thanks for the changes @glindste 
This looks great.
I have a few comments, mostly minor about being explicit about which code is copied and better error messages.

We should be good to merge once we resolve them.
Thanks!

File: app/kubemci/pkg/gcp/healthcheck/healthchecksyncer.go (hunk 12 above)
Comment (glindste) [reply app/kubemci/pkg/gcp/healthcheck/healthchecksyncer.go]: I'm thinking of implementing this as well before I'm done with the PR since that method gets more settings from the probe, like health check interval. Feels a bit half-baked if we go through all the trouble to get the probe and then only use the path.

File: app/kubemci/pkg/gcp/healthcheck/healthchecksyncer_test.go (hunk 9 above)
Comment (glindste) [reply app/kubemci/pkg/gcp/healthcheck/healthchecksyncer_test.go]: At first I had problems with `NewSimpleClientSet` because I didn't know that it adds a catch-all reactor, so if you want to add a reactor that has precedence you need to use `PrependReactor` instead of `AddReactor`. I thought it would be more clear what was going on if we just instantiated Clientset and explicitly set the reactors we need. I think `NewSimpleClientset` is useful when you're just unit-testing one call and you don't care about what call is made, just that it returns a specific object.

File: app/kubemci/pkg/gcp/healthcheck/healthchecksyncer_test.go (hunk 8 above)
Comment (glindste) [reply app/kubemci/pkg/gcp/healthcheck/healthchecksyncer_test.go]: I've changed it but still explicitly set it to `int32`, since `api_v1.ContainerPort` takes an `int32` and there's a TODO in `intstr.FromInt` to change it to take an `int32` as well, but currently we have to cast it to `int`

File: app/kubemci/pkg/gcp/healthcheck/healthchecksyncer.go [hunk 16]
```diff
@@ -231,7 +227,7 @@ func (h *HealthCheckSyncer) desiredHealthCheck(lbName string, port ingressbe.Ser
 	case "HTTP":
 		hc.HttpHealthCheck = &compute.HTTPHealthCheck{
 			Port:        port.Port,
-			RequestPath: path, // TODO(nikhiljindal): Allow customization.
+			RequestPath: path,
```
[the remark this answers is not in the document]
Comment (glindste) [reply app/kubemci/pkg/gcp/healthcheck/healthchecksyncer.go]: Ah, completely missed that!

File: app/kubemci/pkg/gcp/healthcheck/healthchecksyncer.go (hunk 12 above)
Comment (glindste) [reply app/kubemci/pkg/gcp/healthcheck/healthchecksyncer.go]: On second thought I think I'll leave that for later, it might become more straight forward once some refactoring has been done in ingress-gce and I should try to wrap up this PR.

**Comment by glindste** [comment]
@nikhiljindal I think I'm done now, please have a look if I addressed all your comments and if there's anything else I can fix. Thank you for the excellent review comments. Also, would you like me to squash the commits before merging?

**Comment by glindste** [comment]
I saw you'd fixed vet and fmt on master so I rebased and made sure they pass for this branch as well

**Comment by nikhiljindal** [comment]
Yes, please squash the commits.
You can also remove WIP from the PR title and commit message :)
